lib/Target/X86/X86FrameLowering.cpp

```diff
@@ -393,28 +393,152 @@ static bool usesTheStack(const MachineFunction &MF) {
+
+  // Some useful constants
+  const int64_t ThreadEnvironmentStackLimit = 0x10;
+  const int64_t PageSize = 0x1000;
+  const int64_t PageMask = ~(PageSize - 1);
+
+  // Virtual registers we need.
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const TargetRegisterClass *RegClass = &X86::GR64RegClass;
+  const unsigned ZeroReg = MRI.createVirtualRegister(RegClass),
+    CopyReg = MRI.createVirtualRegister(RegClass),
+    TestReg = MRI.createVirtualRegister(RegClass),
+    FinalReg = MRI.createVirtualRegister(RegClass),
+    RoundedReg = MRI.createVirtualRegister(RegClass),
+    LimitReg = MRI.createVirtualRegister(RegClass),
+    JoinReg = MRI.createVirtualRegister(RegClass),
+    ProbeReg = MRI.createVirtualRegister(RegClass);
+
+  // Add code to MBB to check for overflow and set the new target stack pointer
+  // to zero if so.
+  BuildMI(&MBB, DL, TII.get(X86::XOR64rr), ZeroReg)
+    .addReg(ZeroReg, RegState::Undef)
+    .addReg(ZeroReg, RegState::Undef);
+  BuildMI(&MBB, DL, TII.get(X86::MOV64rr), CopyReg).addReg(X86::RSP);
+  BuildMI(&MBB, DL, TII.get(X86::SUB64rr), TestReg)
+    .addReg(CopyReg)
     .addReg(X86::RAX);
+  BuildMI(&MBB, DL, TII.get(X86::CMOVB64rr), FinalReg)
+    .addReg(TestReg)
+    .addReg(ZeroReg);
+
+  // FinalReg now holds final stack pointer value, or zero if allocation would
+  // overflow. Compare against the current stack limit from the thread
+  // environment block.
+  BuildMI(&MBB, DL, TII.get(X86::MOV64rm), LimitReg)
+    .addReg(0)
+    .addImm(1)
+    .addReg(0)
+    .addImm(ThreadEnvironmentStackLimit)
+    .addReg(X86::GS);
+  BuildMI(&MBB, DL, TII.get(X86::CMP64rr)).addReg(FinalReg).addReg(LimitReg);
+  // Jump if the desired stack pointer is at or above the stack limit.
+  BuildMI(&MBB, DL, TII.get(X86::JAE_1)).addMBB(ContinueMBB);
+
+  // Add code to roundMBB to round the final stack pointer to a page boundary.
+  RoundMBB->addLiveIn(X86::RAX);
+  BuildMI(RoundMBB, DL, TII.get(X86::AND64ri32), RoundedReg)
+    .addReg(FinalReg)
+    .addImm(PageMask);
+  BuildMI(RoundMBB, DL, TII.get(X86::JMP_1)).addMBB(LoopMBB);
+
+  // LimitReg now holds the current stack limit, RoundedReg page-rounded
+  // final RSP value. Add code to loopMBB to decrement LimitReg page-by-page
+  // and probe until we reach RoundedReg.
+  LoopMBB->addLiveIn(X86::RAX);
+  BuildMI(LoopMBB, DL, TII.get(X86::PHI), JoinReg)
+    .addReg(LimitReg)
+    .addMBB(RoundMBB)
+    .addReg(ProbeReg)
+    .addMBB(LoopMBB);
+  BuildMI(LoopMBB, DL, TII.get(X86::LEA64r), ProbeReg)
+    .addReg(JoinReg)
+    .addImm(1)
+    .addReg(0)
+    .addImm(-PageSize)
+    .addReg(0);
+  // Probe by storing a byte onto the stack.
+  BuildMI(LoopMBB, DL, TII.get(X86::MOV8mi))
+    .addReg(ProbeReg)
+    .addImm(1)
+    .addReg(0)
+    .addImm(0)
+    .addReg(0)
+    .addImm(0);
+  BuildMI(LoopMBB, DL, TII.get(X86::CMP64rr))
+    .addReg(RoundedReg)
+    .addReg(ProbeReg);
+  BuildMI(LoopMBB, DL, TII.get(X86::JNE_1)).addMBB(LoopMBB);
```

Comment:
As long as this is the canonical approach (i.e. the approach taken by MSVC, etc.), then I'm OK with this.